ao-contracts/curve-bonded-token/src/handlers/bonding_curve.lua

```diff
@@ -0,0 +1,591 @@
+    local reservePID = msg.From
+
+    local sender = msg.Tags.Sender
+
+    local quantityReservesSent = bint(msg.Tags.Quantity)
+
+    -- local slippageTolerance = tonumber(msg.Tags["X-Slippage-Tolerance"]) or 0
+
+    assert(ReachedFundingGoal == false, 'Funding goal has been reached!')
+
+    local tokensToBuy = msg.Tags['X-Token-Quantity']
+    assert(type(tokensToBuy) == 'string', 'Token quantity is required!')
+
+    local tokensToBuyInSubUnits = utils.toSubUnits(tokensToBuy, RepoToken.denomination)
+    assert(bint.__lt(0, tokensToBuyInSubUnits), 'Token quantity must be greater than zero!')
+
+    if (Initialized ~= true or RepoToken == nil or RepoToken.processId == nil) then
+        msg.reply({
+            Action = 'Buy-Tokens-Error',
+            Error = 'Bonding curve not initialized'
+        })
+
+        return
+    end
+
+    -- double call issue
+    local currentSupplyResp = ao.send({ Target = RepoToken.processId, Action = "Total-Supply" }).receive().Data
+    -- local currentSupplyResp = msg.Tags['X-Current-Supply']
+    if (currentSupplyResp == nil) then
+        msg.reply({
+            Action = 'Buy-Tokens-Error',
+            Error = 'Failed to get current supply of curve bonded token'
+        })
+
+        return
+    end
+
+    currentSupplyResp = tostring(currentSupplyResp)
+
+    -- current supply is returned in sub units
+    -- local preAllocation = utils.add(AllocationForLP, AllocationForCreator)
+    local s1 = currentSupplyResp
+    local s2 = utils.add(currentSupplyResp, tokensToBuyInSubUnits);
+    -- Calculate remaining tokens
+    local remainingTokens = utils.subtract(SupplyToSell, currentSupplyResp)
+
+    -- Check if there are enough tokens to sell
+    if bint.__lt(bint(remainingTokens), bint(tokensToBuyInSubUnits)) then
+        msg.reply({
+            Action = 'Buy-Tokens-Error',
+            Remaining = tostring(remainingTokens),
+            TokensToBuy = tostring(tokensToBuyInSubUnits),
+            Error = 'Not enough tokens to sell.'
+        })
+        return
+    end
+
+    local S_exp = bint.ipow(bint(SupplyToSell), bint(EXP_N_PLUS1))
+
+    if bint.__le(S_exp, 0) then
+        msg.reply({
+            Action = 'Buy-Tokens-Error',
+            Error = 'Bonding curve error: S_EXP too low ' .. S_exp
+        })
+
+        return
+    end
+
+    -- Cost = G * [ (s2)^(n+1) - (s1)^(n+1) ] / S^(n+1)
+    local s1_exp = bint.ipow(bint(s1), bint(EXP_N_PLUS1))
+    local s2_exp = bint.ipow(bint(s2), bint(EXP_N_PLUS1))
+
+    local numerator = utils.multiply(FundingGoal, utils.subtract(s2_exp, s1_exp))
+    local cost = utils.divide((numerator), S_exp)
+    LogActivity(msg.Tags['X-Action'], json.encode({ Cost = tostring(cost), AmountSent = tostring(quantityReservesSent) }),
+        "Calculated cost of buying tokens for Reserves sent")
+    if bint.__lt(bint(quantityReservesSent), bint(math.ceil(utils.toNumber(cost)))) then
```

Comment:
Same as before if you need to calculate ceiling in a `bint`, `divide` might return a non-parseable float, plus `tonumber` might truncate